LFS-193: As a user, I can visualize the events recorded for a patient in a timeline

acrowthe opened the pull request:
- [x] Finish code commenting and clean up
- [x] Leveled timeline cards (greyed out for parents, colored dots for children) (LFS-911)
- [x] Move date difference calculation into DateQuestionUtilities, or another utility
- [x] Subject path being prepended to questionniare titles
- [x] "Entered on" dates should be removed
- [x] Bug/regression: the parent link at the top of a child subject chart or timeline is not working (the URL changes correctly, but the page content isn't rerendered)
- [x] The labels of the 2 tabs should be "Chart" (not "PatientChart") and "Timeline"
- [x] Refactoring: the timeline component should be renamed to SubjectTimeline, since it can be applied to any subject, not just to patients
- [x] Time span
- [x] All events occurring on the same date should be in the same bubble

acrowthe committed:
LFS-193: As a user, I can visualize the events recorded for a patient in a timeline
Implement timeline

acrowthe commented:
Currently blocked by and requires LFS-680 (#454), LFS-911(no PR), LFS-795 (#428).
Once these are merged, 6008502 should be removed and TODOs awaiting functionality from these PRs will need to be resolved.
In addition, when rebasing on dev care will need to be taken to avoid overwriting any changes to `Subject.jsx` that should be moved into `PatientChart.jsx` or `SubjectUtilities.jsx`

TODOs awaiting other PRs:
- Entered on dates being displayed wrong (LFS-795)
- Leveled timeline cards (greyed out for parents, colored dots for children) (LFS-911)
- Subject path being prepended to questionniare titles (LFS-911)

sashaandjic commented:
Suggestions:
For events taken from forms such chemo and radio we should consider adding subject reference. In the attached screenshot the patient had 2 radio treatments: 1 for liver cancer and 1 for kidney cancer. In the timeline it is hard to see what cancer got the first radio therapy and when each treatment ended. When hovering over an item in the timeline, it would be really cool if we can mark all the other items that came from the same form, this way the start and end dates will be connected visually too.
![image](https://user-images.githubusercontent.com/23243244/107725671-b74b2380-6cb4-11eb-97ee-820dd7e016a4.png)

sashaandjic commented:
Bugs
SubjectID - wrong format and wrong place:
![image](https://user-images.githubusercontent.com/23243244/107725863-3c363d00-6cb5-11eb-9118-b63973c270b2.png)
Weird date:
![image](https://user-images.githubusercontent.com/23243244/107725882-4b1cef80-6cb5-11eb-88b4-18d5797fd7b2.png)
Broken links:
![image](https://user-images.githubusercontent.com/23243244/107725896-58d27500-6cb5-11eb-9e65-efb463c095e1.png)

marta- commented:
> For events taken from forms such chemo and radio we should consider adding subject reference

Agreed, this was in the specs. If the event is associated with a child subject, prepend "<Subject type> <Subject id>: "  to the form link (e.g. "Tumor 1: Chemotherapy" instead of just "Chemotherapy"). In the specs I also mentioned different colors for the timeline dots belonging to child subjects, which might help visually distinguish between different sources of events.

> When hovering over an item in the timeline, it would be really cool if we can mark all the other items that came from the same form, this way the start and end dates will be connected visually too.

This is a nice to have, but if it doesn't make it into this PR it can certainly be postponed.

acrowthe commented:
> For events taken from forms such chemo and radio we should consider adding subject reference

This will be much easier to implement once LFS-911 gets merged, and is on my list of TODOs before merging. Since This PR is blocked by 680 which is also waiting 911, this can be implemented when 911 gets merged

sashaandjic commented:
Either separate PATIENT from CHART, or just use CHART.
![image](https://user-images.githubusercontent.com/23243244/107826630-3000bd00-6d53-11eb-804e-9070c72eaaf3.png)

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/Subject.jsx:
```diff
@@ -103,7 +133,9 @@ function Subject(props) {
   const [currentSubjectId, setCurrentSubjectId] = useState(id);
+  const [ activeTab, setActiveTab ] = useState(0);
 
+  const tabs = ["PatientChart", "Timeline"]
```
Just "Chart"

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/PatientTimeline.jsx:
```diff
@@ -0,0 +1,214 @@
+ * @param {string} id the identifier of a subject; this is the JCR node name
+ */
+
+function PatientTimeline(props) {
```
Rename to "SubjectTimeline", since it can be applied to any type of subject (Tumor, model organism, ...)

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/PatientChart.jsx:
```diff
@@ -0,0 +1,379 @@
+ * @param {int} maxDisplayed the maximum number of form question/answers to be displayed. defaults to 4
+ */
+
+function PatientChart(props) {
```
Rename to "SubjectChart", since it can be applied to any type of subject (Tumor, model organism, ...)

acrowthe committed:
LFS-193: As a user, I can visualize the events recorded for a patient in a timeline
Implement level and subject titles

sashaandjic commented:
- remove "Entered on 2021-00-01 by user"
- the parent link does not work on a child's chart/timeline (unless you refresh)
![image](https://user-images.githubusercontent.com/23243244/108905331-95736a00-75ed-11eb-965f-62828f74afaf.png)

acrowthe committed:
LFS-193: As a user, I can visualize the events recorded for a patient in a timeline
- Bug fixes
- Add date duration to timeline

acrowthe commented:
Added date difference
![image](https://user-images.githubusercontent.com/20056751/109556588-0232ac80-7aa5-11eb-99f3-82074dc1436f.png)

sashaandjic commented:
- Put space between Patient and CHART
- previously we were showing only 4 form entries on the chart
![image](https://user-images.githubusercontent.com/23243244/109575948-d6252480-7ac0-11eb-85c7-1a82822ffce7.png)

sashaandjic commented:
pressing "+" on Patient chart/Timeline does not work
![image](https://user-images.githubusercontent.com/23243244/109576294-83983800-7ac1-11eb-804e-438a0bd00332.png)

sashaandjic commented:
I do not see any changes. What am I doing wrong? 
![image](https://user-images.githubusercontent.com/23243244/109579528-1091c000-7ac7-11eb-9187-3d096f2cc1e4.png)

sashaandjic commented:
This is also strange
![image](https://user-images.githubusercontent.com/23243244/109588398-a97c0780-7ad6-11eb-80c7-6e74f4a0b5c9.png)

acrowthe commented:
@sashaandjic All 4 problems look like problems I was having with an older version of this branch: I suspect my push didn't work properly, since I was trying to push from a local backup branch to this branch. I'll let you know when I have the right version pushed

acrowthe commented:
Ok, the correct version has now been uploaded

sashaandjic commented:
Looking really good!
![screencapture-localhost-1930-content-html-Subjects-d9b2807c-0947-4881-a69a-ec9e20bd41f5-2021-03-02-14_11_07](https://user-images.githubusercontent.com/23243244/109702193-eccf8800-7b61-11eb-9b0e-872de415e5a0.png)

acrowthe committed:
LFS-193: As a user, I can visualize the events recorded for a patient in a timeline
Fix rebase conflicts

acrowthe committed:
LFS-193: As a user, I can visualize the events recorded for a patient in a timeline
Finish implementing parent timeline entries
Implement merged timeline cards with multiple answers on the same date

acrowthe committed:
LFS-193: As a user, I can visualize the events recorded for a patient in a timeline
Begin code clean up

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/QuestionnaireStyle.jsx:
```diff
@@ -233,6 +233,31 @@ const questionnaireStyle = theme => ({
     },
+    timelinePaper: {
+        padding: "8px 16px",
+        marginBottom: "40px"
```
The bubbles need to be pulled away a bit from the "duration" circles.
```suggestion
        padding: theme.spacing(1,2),
        marginBottom: theme.spacing(0, 1, 5),
```

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/QuestionnaireStyle.jsx:
```diff
@@ -233,6 +233,31 @@ const questionnaireStyle = theme => ({
+             top: 0,
+             right: 0,
+        },
+    },
```
This should be removed

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/QuestionnaireStyle.jsx:
```diff
@@ -472,6 +497,20 @@ const questionnaireStyle = theme => ({
+        borderRadius: "50%",
+        display: "flex",
+        alignItems: "center",
+        justifyContent: "center"
```
Timeline-related styles should be grouped together

marta- replied:
I think these would look better filled, lighter gray, and shaped more like "lumps":

![image](https://user-images.githubusercontent.com/651980/111040940-b7e5ef80-8403-11eb-8933-cebf1932ac74.png)

Here's what I did din the browser dev tools to get the look:

```
.circle:before {
    content: "";
    display: block;
    border-bottom: 12px solid #ddd;
    border-left: 12px solid transparent;
    border-right: 12px solid transparent;
    position: absolute;
    top: -10px;
}
.circle:after {
    content: "";
    display: block;
    border-top: 12px solid #ddd;
    border-left: 12px solid transparent;
    border-right: 12px solid transparent;
    position: absolute;
    bottom: -10px;
}
```

Also in the mockup above, the timeline lines are the same light gray, and the dates are midnight blue, closer to the circle, and aligned with it vertically.

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/Subject.jsx:
```diff
@@ -592,4 +669,4 @@ Subject.defaultProps = {
 }
 
-export default withStyles(QuestionnaireStyle)(Subject);
+export default withStyles(QuestionnaireStyle)(Subject);
```
Add new line back

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/QuestionnaireStyle.jsx:
```diff
@@ -233,6 +233,31 @@ const questionnaireStyle = theme => ({
+        marginBottom: "40px"
+    },
+    timelineAncestor: {
+        opacity: 0.3
```
We should probably bring them to full opacity on hover:
```suggestion
        opacity: 0.3,
        "&:hover": {
          opacity: 1,
        },
```

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/SubjectTimeline.jsx:
```diff
@@ -0,0 +1,325 @@
+      <TimelineOppositeContent>
+        <Typography color="textSecondary">{dateText}</Typography>
+      </TimelineOppositeContent>
+      <TimelineSeparator className={dateEntry.level < 0 ? classes.timelineAncestor : null}>
```
In the screenshot below, the separator and duration bubble right under **End (Chemo)** should also be faded.

![image](https://user-images.githubusercontent.com/651980/111041377-e49b0680-8405-11eb-91a6-200f40025556.png)

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/Subject.jsx:
```diff
@@ -131,18 +135,41 @@ function Subject(props) {
+          </Tabs>
+        }
+        {
+          activeTab === tabs.indexOf("Chart")
```
It would be nice to be able to specify the active tab in the URL, e.g. in the anchor.

acrowthe replied:
Added an active tab to the anchor and set up browser history so refreshing the page opens the same tab. Shouldn't cause any issues with the back button.

marta- reviewed (commented):
Looks great

acrowthe committed:
LFS-193: As a user, I can visualize the events recorded for a patient in a timeline
Imporve styling and add chart vs. timeline anchor

acrowthe committed:
LFS-193: As a user, I can visualize the events recorded for a patient in a timeline
Fininsh code cleanup

acrowthe commented:
Updated, should implement all requested style changes:
![image](https://user-images.githubusercontent.com/20056751/111338673-7cd6fc80-864d-11eb-94cf-e0ff60443c31.png)

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/QuestionnaireStyle.jsx:
```diff
@@ -233,6 +233,74 @@ const questionnaireStyle = theme => ({
+        },
+    },
+    timelineDate: {
+        marginTop: "-5px",
```
We might be able to obtain the same effect without hardcoding pixel sizes, like this:
```suggestion
        lineHeight: 1em,
```

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/QuestionnaireStyle.jsx:
```diff
@@ -233,30 +233,73 @@ const questionnaireStyle = theme => ({
+    },
     timelinePaper: {
-        padding: "8px 16px",
-        marginBottom: "40px"
```
I think removing this bottom margin makes the timeline harder to read. Each `timelinePaper` ends up to close to the date right below it, so it's not immediately obvious with which date it is actually associated. We can reduce it a bit from 40px to 32 or even 24 (`theme.spacing(3)`), but I wouldn't remove it completely.

acrowthe replied:
Adjusted padding back to 24px. It looks reasonable to me, but it's easy to add more padding if desired

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/SubjectTimeline.jsx:
```diff
@@ -0,0 +1,370 @@
+    return <CircularProgress/>
+  }
+
+  return <Timeline align="alternate">
```
Let's add a max-width to the timeline and center it horizontally, to distribute the negative space around it rather than stretching the timeline entries:

![image (30)](https://user-images.githubusercontent.com/651980/111366141-85d5c700-8669-11eb-85f7-c9fc4b9ca9a5.png)

acrowthe committed:
LFS-193: As a user, I can visualize the events recorded for a patient in a timeline
Adjust vertical spacing and implement max timeline width

acrowthe commented:
Spacing fixed and max width implemented
![image](https://user-images.githubusercontent.com/20056751/111383643-e02d5280-867e-11eb-933f-d3b520824e1c.png)

acrowthe commented:
Spacing fixed and max width implemented
![image](https://user-images.githubusercontent.com/20056751/111383643-e02d5280-867e-11eb-933f-d3b520824e1c.png)

sashaandjic reviewed (approved):
tested

OrdiNeu commented on modules/data-entry/src/main/frontend/src/questionnaire/Subject.jsx:
```diff
@@ -223,45 +234,115 @@ function SubjectContainer(props) {
+  // such as authorship and versioning information.
+  // Once the data arrives from the server, it will be stored in the `data` state variable.
+  let fetchSubjectData = () => {
+    fetchWithReLogin(globalLoginDisplay, `/Subjects/${id}.progeny.deep.json`)
```
```suggestion
    fetchWithReLogin(globalLoginDisplay, `/Subjects/${id}.deep.json`)
```
The progeny selector is no longer used as of LFS-911

OrdiNeu commented on modules/data-entry/src/main/frontend/src/questionnaire/Subject.jsx:
```diff
@@ -223,45 +234,115 @@ function SubjectContainer(props) {
+    );
+  }
+
+  let identifier = subject?.data?.identifier ? subject.data.identifier : id;
```
```suggestion
  let identifier = subject?.data?.identifier || id;
```

OrdiNeu commented on modules/data-entry/src/main/frontend/src/questionnaire/Subject.jsx:
```diff
@@ -223,45 +234,115 @@ function SubjectContainer(props) {
+                 buttonClass={classes.subjectHeaderButton}
+                 size={"large"}
+               />
+  let parentDetails = subject?.data && subject.data['parents'] && getHierarchy(subject.data['parents']);
```
```suggestion
  let parentDetails = subject?.data?.['parents'] && getHierarchy(subject.data['parents']);
```

OrdiNeu commented on modules/data-entry/src/main/frontend/src/questionnaire/Subject.jsx:
```diff
@@ -223,45 +234,115 @@ function SubjectContainer(props) {
+  let identifier = subject?.data?.identifier ? subject.data.identifier : id;
+  let label = subject?.data?.type?.label || "Subject";
+  let title = `${label} ${identifier}`;
+  let path = subject?.data ? subject.data["@path"] : "/Subjects/" + id;
```
```suggestion
  let path = subject?.data?.["@path"] : "/Subjects/" + id;
```

OrdiNeu commented on modules/data-entry/src/main/frontend/src/questionnaire/Subject.jsx:
```diff
@@ -223,45 +234,115 @@ function SubjectContainer(props) {
       <Typography variant="h2">{title}{action}</Typography>
       {
-        data && data['jcr:createdBy'] && data['jcr:created'] ?
+        subject?.data && subject.data['jcr:createdBy'] && subject.data['jcr:created'] ?
```
```suggestion
        subject?.data?.['jcr:createdBy'] && subject?.data?.['jcr:created'] &&
```
Along with removing the `: ""` on line 342

OrdiNeu replied:
In addition, is there a reason we check for both? I.e. is there a case where only one of the two exists?

OrdiNeu commented on modules/data-entry/src/main/frontend/src/questionnaire/SubjectTimeline.jsx:
```diff
@@ -0,0 +1,370 @@
+ * @param {string} id the identifier of a subject; this is the JCR node name
+ */
+function SubjectTimeline(props) {
+  let { id, classes, pageSize, subject } = props;
```
```suggestion
  let { id, classes, subject } = props;
```
Unused

OrdiNeu reviewed (commented):
![image](https://user-images.githubusercontent.com/4656440/111532565-2b596b00-873c-11eb-9452-0072599c8c1a.png)

Not sure if this is a bug, but the dates don't seem to be rounded when calculating the time distance between things. In this case, I added a `hh:mm` to StartDate and made it begin at 4:23. The distance between the two dates (the 3rd of March and the 10th) is then recorded as 6 days, rather than 7. Adding another event on the same day (March 3rd) then fixes this, as below:

![image](https://user-images.githubusercontent.com/4656440/111533050-bcc8dd00-873c-11eb-89ff-e8210a50cab1.png)

acrowthe commented:
> Not sure if this is a bug, but the dates don't seem to be rounded when calculating the time distance between things. In this case, I added a `hh:mm` to StartDate and made it begin at 4:23. The distance between the two dates (the 3rd of March and the 10th) is then recorded as 6 days, rather than 7. Adding another event on the same day (March 3rd) then fixes this, as below:

Rounding isn't performed, for simplicity. The way it works is take the difference, then use the longest present time period:
In your case of the 3rd at 04:00 and the 10th (implicitly at 00:00), this gives a difference of 6days, 20 hours, or 6d. In the case of grouped events, the earliest date in the group is used for comparison, so the new diagnosis date (3rd at 00:00) is used, resulting in 7 days exactly.

I could fix this if desired by completely ignoring hours and minutes.

acrowthe committed:
LFS-193: As a user, I can visualize the events recorded for a patient in a timeline
Clean up Subject conditional statements
Fix unexpected date differences when comparing dateTimes by ignoring hh:mm:ss

acrowthe commented:
Hours, minutes and seconds are now ignored by the date difference calculation, so the reported inconsistency should be resolved

sashaandjic commented:
How precise we want to be here? The circled items are 10 years apart, however by counting numbers shown in the "bubbles" it appears to be 9. Should we make a disclaimer that the number of years are rounded down ( e.g. 6y8m is shown as 6y)
![image](https://user-images.githubusercontent.com/23243244/112230547-614e9180-8c0b-11eb-891c-78e7bacd149b.png)

sashaandjic reviewed (approved):
tested - I think it is good as is; my comment about the precision may be taken into consideration at a later date

marta- commented:
> How precise we want to be here? The circled items are 10 years apart, however by counting numbers shown in the "bubbles" it appears to be 9. Should we make a disclaimer that the number of years are rounded down ( e.g. 6y8m is shown as 6y)

We should round up from N.5, that would reduce some inconsistencies (but not eliminate them completely). I think people will figure out there's some rounding error here, no need for a disclaimer. Alternatively (or additionally?) we could show a more precise duration on hover (e.g. 6y8m3d).

acrowthe committed:
LFS-193: As a user, I can visualize the events recorded for a patient in a timeline
Add rounding and full difference tooltip to timeline bubbles

acrowthe commented:
Added rounding. If month >6 or date > 15, round up. Date rounding will be incorrect for February, but should be close enough.
Also added the full difference to a tooltip

![image](https://user-images.githubusercontent.com/20056751/112340391-dfa44580-8c96-11eb-93d4-eccded403782.png)

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/DateQuestionUtilities.jsx:
```diff
@@ -158,19 +158,29 @@ export default class DateQuestionUtilities {
+        diff.push(value);
+        if (value > 0) {
+          nextDate = nextDate.subtract(value, division);
+          result.long += value + division.charAt(0);
```
Looks great. Can we add a space between divisions? I think '1y 5m 2d' would be easier to read

sashaandjic commented:
Looks awesome. Waiting for the final change that @marta- suggested (space between y m and d in the tooltip) before approving.

acrowthe committed:
LFS-193: As a user, I can visualize the events recorded for a patient in a timeline
Add spaces to date difference tooltip

acrowthe commented:
Added spaces:
![image](https://user-images.githubusercontent.com/20056751/112482207-2dc85000-8d4e-11eb-83de-8379423dfdfb.png)
![image](https://user-images.githubusercontent.com/20056751/112482224-31f46d80-8d4e-11eb-96b2-99aa8e084996.png)

sashaandjic commented:
Not sure what is happening here.
![image](https://user-images.githubusercontent.com/23243244/112577837-8c7fdf00-8dcb-11eb-96af-e02b90bde635.png)
![image](https://user-images.githubusercontent.com/23243244/112577856-9570b080-8dcb-11eb-9b3f-040ee9a1f6fb.png)
![image](https://user-images.githubusercontent.com/23243244/112577869-9c97be80-8dcb-11eb-9289-3a4447126a05.png)

sashaandjic commented:
Somatic Variants files are showing under Radiation form:
![image](https://user-images.githubusercontent.com/23243244/112579601-5ee76580-8dcc-11eb-9ab3-6b072a9849c5.png)

marta- commented:
> Not sure what is happening here.

Could the date shift be introduced by stripping hours, minutes, seconds, and timezone?

marta- commented:
> Somatic Variants files are showing under Radiation form:

@acrowthe Is every bubble showing the next questions in the same form or the next questions in the entire queue?

acrowthe commented:
> Could the date shift be introduced by stripping hours, minutes, seconds, and timezone?

That's my first guess, especially since Sasha put these comments up around midnight. Not happening right now to my dev environment, so I'm going to play around with my VM's time/timezone to try and reproduce this.

> @acrowthe Is every bubble showing the next questions in the same form or the next questions in the entire queue?

It should just be the questions in the same form. but seems like it is using the entire queue. This will be an easy fix

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaire/DateQuestionUtilities.jsx:
```diff
@@ -155,4 +155,34 @@ export default class DateQuestionUtilities {
+    // Compute the displayed difference
+    let result = {long:""}
+    if (firstDate && secondDate) {
+      let currentDate = this.amendMoment(firstDate, "yyyy-MM-dd");
```
I would name this `baseDate` instead of `currentDate`, because `current` means `now`.

marta- commented:
> > Not sure what is happening here.
> 
> Could the date shift be introduced by stripping hours, minutes, seconds, and timezone?

It was determined that the cause was the testing environment setup, so this issue doesn't need to be addressed by this PR.

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaire/Subject.jsx:
```diff
@@ -33,16 +33,18 @@ import {
-const ENTRY_TYPES = QUESTION_TYPES.concat(SECTION_TYPES);
+export const QUESTION_TYPES = ["lfs:Question"];
+export const SECTION_TYPES = ["lfs:Section"];
+export const ENTRY_TYPES = QUESTION_TYPES.concat(SECTION_TYPES);
```
These are already defined (and exported) in `FormEntry.jsx`, you should import them from there.

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaire/Subject.jsx:
```diff
@@ -100,15 +102,20 @@ export function getTextHierarchy (node, withType = false) {
+  const [ currentSubjectId, setCurrentSubjectId ] = useState(id);
+  const [ activeTab, setActiveTab ] = useState(0);
 
+  const tabs = ["Chart", "Timeline"]
```
We need to make this extensible soon.

acrowthe replied:
Agreed. I've added a comment detailing what needs to happend to do so, but didn't want to touch the patient chart more than is required in this PR to avoid merge issues

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaire/Subject.jsx:
```diff
@@ -124,25 +131,51 @@ function Subject(props) {
+
   return (
     <React.Fragment>
-      <div className={classes.mainPageAction}>
```
Why is this not used anymore?

marta- replied:
NewFormDialog already renders the + button with the right styling. This wrapper is an ommission from when the (+) buttons became reusable components.

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaire/Subject.jsx:
```diff
@@ -124,25 +131,51 @@ function Subject(props) {
 
-/**
- * Component that recursively gets and displays the selected subject and its related SubjectTypes
- */
```
Why remove the doc?

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaire/Subject.jsx:
```diff
@@ -223,45 +234,115 @@ function SubjectContainer(props) {
-function SubjectHeader (props) {
-  let {data, classes, title, action} = props;
-  let parentDetails = data && data['parents'] && getHierarchy(data['parents']);
+function SubjectHeader(props) {
```
I would move this above the `SubjectContainer`.

marta- replied:
Preserving the original order of components in the file makes reviewing easier.

sdumitriu replied:
Kind of, but other than the component name, this whole block is completely new and could be inserted anywhere.

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaire/Subject.jsx:
```diff
@@ -223,45 +234,115 @@ function SubjectContainer(props) {
+  };
+
+  // Fetch this Subject's data
+  useEffect(() => {
```
I think this may trigger react warnings about hooks being called out of order because of the early return above. For a component, the exact same hooks must be invoked every time. It should be OK to move the return after this effect.

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaire/Subject.jsx:
```diff
@@ -319,7 +396,7 @@ function SubjectMemberInternal (props) {
   useEffect(() => {
     fetchTableData();
-  }, []);
+  }, [data]);
```
Complex objects should not be used as dependencies, but I think in this case, since we don't modify the data after the initial fetch returns, the exact same object should be referenced.

acrowthe replied:
Switched to data["jcr:uuid"]

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaire/SubjectTimeline.jsx:
```diff
@@ -0,0 +1,386 @@
+function DateAnswerDisplay(classes, questionData, index, length, rootLevel) {
+  let questionTitle = questionData.questionText;
+  // Strip unwanted strings from the question title
+  for (const str of STRIPPED_STRINGS) {
+    let dateIndex = questionTitle.toLowerCase().indexOf(str);
+    if (dateIndex > -1) {
+      questionTitle = questionTitle.substring(0,dateIndex).concat(questionTitle.substring(dateIndex + str.length));
+    }
+  }
```
I think it would be simpler to do this with a regexp replace, and I think it would be even safer to use regexps for that to enforce some bounds, since right now a question named "Patient updated on" becomes "Patient upd on". This would be safer:

```
[/^date of +/i, / +date$/i]
```

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaire/SubjectTimeline.jsx:
```diff
@@ -0,0 +1,386 @@
+      .catch(response => handleError(response, ""))
+  }
+
+  // Get the full set of data for all forms provided
```
Why not request `/Subjects/<subject id>.data.deep.json`? This would give all the subject's forms. Of course, the subject fetched should be the topmost subject, not the current one. I think it would reduce a lot of code complexity and network traffic. Would that work?

acrowthe replied:
This has been implemented. It helps, as the subject, forms and answers can be fetched in one call. It only fetches the current subject's forms though, so traversing the tree of subjects is still required.

acrowthe committed:
LFS-193: As a user, I can visualize the events recorded for a patient in a timeline
Fix questions from multiple forms showing up on a single timeline date entry

sashaandjic commented:
Timeline is sometimes showing patient details sometimes not. In the sample below, I created several completed forms. Not a single info from the demographic form was shown. We should either show the details in the timeline, or not show them.
<img width="960" alt="TIMELINE" src="https://user-images.githubusercontent.com/23243244/112766697-e1487300-8fe0-11eb-904f-58fb4163b3b6.png">
![CHART](https://user-images.githubusercontent.com/23243244/112766698-e1487300-8fe0-11eb-9c81-a9b052b41ba1.png)

sashaandjic commented:
Inconsistent labeling for Radio and Chemo `start `and `end`:
![image](https://user-images.githubusercontent.com/23243244/112853995-684a2980-907b-11eb-973f-58d86254742f.png)

acrowthe commented:
> Inconsistent labeling for Radio and Chemo start and end:

@sashaandjic That is due to inconsistencies in the base forms. As the label is determined by removing "date" from the question title, that inconsistency is reflected:
![image](https://user-images.githubusercontent.com/20056751/112857836-45217900-907f-11eb-91cf-bae5ec4b9798.png)
![image](https://user-images.githubusercontent.com/20056751/112857908-58344900-907f-11eb-9c5a-9c38c9ad7f21.png)

acrowthe committed:
LFS-193: As a user, I can visualize the events recorded for a patient in a timeline
Reduce number of fetches using .data processor and address code comments
Fix followup questions not always showing up

acrowthe commented:
> Timeline is sometimes showing patient details sometimes not

Should be resolved now. The wrong condition was being used to determine which questions were from the current form

sashaandjic commented:
Looks really great.
![LFS-193](https://user-images.githubusercontent.com/23243244/113342869-07318880-92fd-11eb-8abf-c0d73b635ab5.png)

sashaandjic reviewed (approved):
tested

acrowthe committed:
LFS-193: As a user, I can visualize the events recorded for a patient in a timeline
Pass classes object to required function

acrowthe commented:
Rebased on top of dev

sashaandjic reviewed (approved):
tested

sdumitriu reviewed (approved):
Displaying the "followup" questions in the item box has some clunky results, like in Demographics / Surveillance Details, "Surveyed until" shows as details "Age at 1st malignant Dx" and "Inherited from", which have nothing to do with the surveillance. I think a good fix for that is to:

- not flatten the answers list too soon, so that we don't see answers from outside the current section
- properly split our forms into sections, so that every date question doesn't have unrelated questions following it

But this can wait.
